[flang] Updating drivers to create data layout before semantics

jeanPerier opened the pull request:
Preliminary patch to change lowering/code generation to use llvm::DataLayout information instead of generating "sizeof" GEP (see https://github.com/llvm/llvm-project/issues/71507).

Fortran Semantic analysis needs to know about the target type size and alignment to deal with common blocks, and intrinsics like C_SIZEOF/TRANSFER. This information should be obtained from the llvm::DataLayout so that it is consistent during the whole compilation flow.

This change is changing flang-new and bbc drivers to:
1. Create the llvm::TargetMachine so that the data layout of the target can be obtained before semantics.
2. Sharing bbc/flang-new set-up of the SemanticConstext.targetCharateristics from the llvm::TargetMachine. For now, the actual part that set-up the Fortran type size and alignment from the llvm::DataLayout is left TODO so that this change is mostly an NFC impacting the drivers.
3. Let the lowering bridge set-up the mlir::Module datalayout attributes since it is doing it for the target attribute, and that allows the llvm data layout information to be available during lowering.

For flang-new, the changes are code shuffling: the `llvm::TargetMachine` instance is moved to `CompilerInvocation` class so that it can be used to set-up the semantic contexts. `setMLIRDataLayout` is moved to `flang/Optimizer/Support/DataLayout.h` (it will need to be used from codegen pass for fir-opt target independent testing.)), and the code setting-up semantics targetCharacteristics is moved to `Tools/TargetSetup.h` so that it can be shared with bbc.

As a consequence, LLVM targets must be registered when running semantics, and it is not possible to run semantics for a target that is not registered with the -triple option (hence the power pc specific modules can only be built if the PowerPC target is available, @DanielCChen and @kkwli , I think I got it right, but you may want to double check the `tools/f18/CMakeLists.txt` change).

jeanPerier committed:
[flang] Updating drivers to create data layout before semantics

Preliminary patch to change lowering/code generation to use
llvm::DataLayout information instead of generating "sizeof" GEP (see
https://github.com/llvm/llvm-project/issues/71507).

Fortran Semantic analysis needs to know about the target type size and
alignment to deal with common blocks, and intrinsics like
C_SIZEOF/TRANSFER. This information should be obtained from the
llvm::DataLayout so that it is consistent during the whole compilation
flow.

This change is changing flang-new and bbc drivers to:
1. Create the llvm::TargetMachine so that the data layout of the target
   can be obtained before semantics.
2. Sharing bbc/flang-new set-up of the SemanticConstext.targetCharateristics
   from the llvm::TargetMachine. For now, the actual part that set-up
   the Fortran type size and alignment from the llvm::DataLayout is left
   TODO so that this change is mostly an NFC impacting the drivers.
3. Let the lowering bridge set-up the mlir::Module datalayout attributes
   since it is doing it for the target attribute, and that allows the
   llvm data layout information to be available during lowering.

As a consequence, LLVM targets must be registered when running
semantics, and it is not possible to run semantics for a target that
is not registered with the -triple option (hence the power pc specific
modules can only be built if the PowerPC target is available).

kiranchandramohan commented:
CI seems to be failing while building the compiler.
```
/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h5ngp-1/llvm-project/github-pull-requests/flang/lib/Frontend/CompilerInvocation.cpp:1505:10: error: 'tm' does not refer to a value
--
  | assert(tm && "Failed to create TargetMachine");
  | ^
  | /usr/include/x86_64-linux-gnu/bits/types/struct_tm.h:7:8: note: declared here
  | struct tm
```

```

Building CXX object tools\flang\lib\Frontend\CMakeFiles\obj.flangFrontend.dir\CompilerInvocation.cpp.obj
--
  | FAILED: tools/flang/lib/Frontend/CMakeFiles/obj.flangFrontend.dir/CompilerInvocation.cpp.obj
  | sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\flang\lib\Frontend -IC:\ws\src\flang\lib\Frontend -IC:\ws\src\flang\include -Itools\flang\include -Iinclude -IC:\ws\src\llvm\include -IC:\ws\src\llvm\..\mlir\include -Itools\mlir\include -Itools\clang\include -IC:\ws\src\llvm\..\clang\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\flang\lib\Frontend\CMakeFiles\obj.flangFrontend.dir\CompilerInvocation.cpp.obj /Fdtools\flang\lib\Frontend\CMakeFiles\obj.flangFrontend.dir\ /FS -c C:\ws\src\flang\lib\Frontend\CompilerInvocation.cpp
  | C:\ws\src\flang\lib\Frontend\CompilerInvocation.cpp(1505): error C2143: syntax error: missing ')' before 'string'
  | C:\ws\src\flang\lib\Frontend\CompilerInvocation.cpp(1505): error C2678: binary '!': no operator found which takes a left-hand operand of type 'tm' (or there is no acceptable conversion)
  | C:\ws\src\flang\lib\Frontend\CompilerInvocation.cpp(1505): note: could be 'built-in C++ operator!(bool)'
  | C:\ws\src\flang\lib\Frontend\CompilerInvocation.cpp(1505): note: while trying to match the argument list '(tm)'
  | C:\ws\src\flang\lib\Frontend\CompilerInvocation.cpp(1505): error C2088: '!': illegal for struct
  | C:\ws\src\flang\lib\Frontend\CompilerInvocation.cpp(1505): error C2059: syntax error: ')'
```

kiranchandramohan commented on flang/test/Lower/bbc-host-datalayout.f90:
```diff
@@ -0,0 +1,9 @@
+! Test bbc set-up of the target data layout from the host.
+! RUN: bbc %s -o - | FileCheck %s
```
Assuming there are existing tests for the flang-new driver and new ones are not required.

jeanPerier replied:
That is right, flang-new changes are NFC aside from the timing when the TargetMachine is created.
This is already tested here:
- https://github.com/llvm/llvm-project/blob/820b3583c9681a1969814ac3b04a6fe41b87d079/flang/test/Driver/emit-llvm.f90#L9
- https://github.com/llvm/llvm-project/blob/820b3583c9681a1969814ac3b04a6fe41b87d079/flang/test/Driver/emit-mlir.f90#L13

kiranchandramohan reviewed (commented):
> Fortran Semantic analysis needs to know about the target type size and alignment to deal with common blocks, and intrinsics like C_SIZEOF/TRANSFER. This information should be obtained from the llvm::DataLayout so that it is consistent during the whole compilation flow.

Is it possible to add a semantics test for this?

jeanPerier commented:
> CI seems to be failing while building the compiler

Thanks, should be fixed.

> Is it possible to add a semantics test for this?

No because the functional change to fill the SemanticsContext.targetCharacteristics from the TargetMachine is not part of this change, it is left TODO here: https://github.com/llvm/llvm-project/pull/73301/files#diff-74352fcca4246990c7d9ce870cefdb335412f10f045e0175ef242cbc79564f2eR37

The current patch only ensures the TargetMachine is created before setting up the SemanticsContext.targetCharacteristics in the drivers and that the llvm::DataLayout and its MLIR equivalent are always available during lowering/codegen.

kkwli commented on flang/tools/f18/CMakeLists.txt:
```diff
@@ -49,9 +49,14 @@ if (NOT CMAKE_CROSSCOMPILING)
+      else()
+        # Do not compile PPC module if the target is not available.
+        continue()
+      endif()
```
Yes. These modules are only built for PPC platform.

banach-space commented:
I've quickly scanned through and have one high level comment - would it be possible to keep the logic related to `TargetMachine` in `FrontendAction`?

This change is driven by the need to access `TargetMachine` during the semantic analysis. So, it makes to move `TargetMachine` from `CodeGenAction` somewhere higher up in the class hierarchy. As in, it's no longer a detail specific to `CodeGenAction`. However, this should be possible by moving it to the base class of `CodeGenAction` (shared with other actions that run semantic checks) rather than all the way to `CompilerInvocation`. 

One reason for this ask would be to make these changes less intrusive. But that's secondary. The main reason would be to stick to the original design principles behind these classes. `CompilerInvocation` merely represents ... the compiler invocation with all the options requested by the user and/or dictated by the environment (and a few other details). While `TargetMachine` can be configured based on some input from `CompilerInvocation`, it isn't really needed until the compiler is actually run - and that's represented/encapsulated by `FrontendAction` and its various sub-classes (e.g. `CodeGenAction`).

I think that my suggestion could be implemented by extracting `TargetMachine` logic out of `setUpTargetCharacteristics`? Hopefully that wouldn't be too much work. 

@jeanPerier would that still work for you?

banach-space commented on flang/include/flang/Tools/TargetSetup.h:
```diff
@@ -0,0 +1,63 @@
+/// contain all the CPU feature information to get optimized assembly generation
+/// from LLVM IR. Drivers that needs to generate assembly from LLVM IR should
+/// create a target machine according to their specific options.
+[[maybe_unused]] inline static std::unique_ptr<llvm::TargetMachine>
+createTargetMachine(llvm::StringRef targetTriple, std::string &error) {
+  std::string triple{targetTriple};
+  if (triple.empty())
+    triple = llvm::sys::getDefaultTargetTriple();
+
+  const llvm::Target *theTarget =
+      llvm::TargetRegistry::lookupTarget(triple, error);
+  if (!theTarget)
+    return nullptr;
+  return std::unique_ptr<llvm::TargetMachine>{
+      theTarget->createTargetMachine(triple, /*CPU=*/"",
+          /*Features=*/"", llvm::TargetOptions(),
+          /*Reloc::Model=*/std::nullopt)};
+}
```
Given: https://github.com/llvm/llvm-project/blob/9ba74c2ee044381544c8e1e736c4c85a1a9d56bf/llvm/include/llvm/MC/TargetRegistry.h#L484-L500

this method shouldn't really be called `createTargetMachine` (to avoid confusion). If a specialisation of the original method is required, then IMHO it should live in LLVM. 

Also, from what I can see this is only needed by `bbc`, so I suggest moving it there. "Tools" based on LLVM Flang can leverage the APIs exposed by the driver logic that will handle this sort of configuration automatically.

jeanPerier commented:
> I think that my suggestion could be implemented by extracting `TargetMachine` logic out of `setUpTargetCharacteristics`?

I want a single place where the evaluate::TargetCharacteristics is being set-up, and it will need an llvm::DataLayout, which can only be obtained from an llvm::TargetMachine.

The other solution is to move the SemanticContext outside of CompilerInvocation, but I am not sure of the implications. Would that be possible?

jeanPerier committed:
move createTargetMachine to bbc

vzakhari commented on flang/include/flang/Optimizer/Support/DataLayout.h:
```diff
@@ -0,0 +1,39 @@
+/// These attributes are replaced if they were already set.
+void setMLIRDataLayout(mlir::ModuleOp mlirModule, const llvm::DataLayout &dl);
+
+/// Create an mlir::DataLayoutSpecInterface from the llvm>data_layout attribute
```
nit:
```suggestion
/// Create an mlir::DataLayoutSpecInterface from the llvm.data_layout attribute
```

banach-space commented:
> > > I think that my suggestion could be implemented by extracting `TargetMachine` logic out of `setUpTargetCharacteristics`?
> > 
> > 
> > I want a single place where the evaluate::TargetCharacteristics is being set-up, and it will need an llvm::DataLayout, which can only be obtained from an llvm::TargetMachine.
> > The other solution is to move the SemanticContext outside of CompilerInvocation, but I am not sure of the implications. Would that be possible?
> 
> @banach-space, alternatively, would it be OK to move the llvm::TargetMachine into the `CompilerInstance` class where the call to set-up the SemanticContext is done?

Interesting that you mention that -  `semanticContext` is currently a member variable of `CompilarInvocation` rather than `CompilerInstance`: https://github.com/llvm/llvm-project/blob/79b03306af5c11d354fa90db8bfd7818cd811ef5/flang/include/flang/Frontend/CompilerInvocation.h#L90-L91

That's not where I was expecting to find it :)

It sounds like things would be more seamless for both of us if I moved `semanticContext` to `CompilerInstance`. I have a prototype patch that does that - would you be OK  rebasing your change on top of that if I send it for review?

jeanPerier committed:
Fix typo in comment

Co-authored-by: Slava Zakharin <szakharin@nvidia.com>

jeanPerier commented:
> would you be OK rebasing your change on top of that if I send it for review?

@banach-space, I just rebased on a main with your patch and moved the target machine into `CompilerInstance`.

banach-space commented:
The driver changes LGTM!

The stuff that's been added to CompilerInstance.cpp (e.g. `getExplicitAndImplicitAMDGPUTargetFeatures`) should be extracted to a different file and probably encapsulated with some fancy "Target info" class. But that's tangential to this change.

Thank you for addressing my comments 🙏🏻

kiranchandramohan reviewed (approved):
LG. Thanks Jean.
